lib/strdup.c

```diff
@@ -104,18 +104,14 @@ void *Curl_memdup(const void *src, size_t length)
  * Curl_strndup(source, length)
  *
  * Copies the 'source' string to a newly allocated buffer (that is returned).
- * Copies not more than 'length' bytes (up to a null terminator) then adds a
- * null terminator.
+ * Copies 'length' bytes then adds a null terminator.
  *
  * Returns the new pointer or NULL on failure.
  *
  ***************************************************************************/
 void *Curl_strndup(const char *src, size_t length)
 {
-  char *buf = memchr(src, '\0', length);
-  if(buf)
-    length = buf - src;
-  buf = malloc(length + 1);
+  char *buf = malloc(length + 1);
   if(!buf)
     return NULL;
```

Comment:
> I want to remove the memchr stuff again from strndup() because I think it potentially breaks several of these strndup cases.

How is it breaking? If we don't check if the length of src is shorter than the passed in length then the memcpy can read past the end of src. If the full passed in length needs to be allocated then it could be adjusted after the allocation but memchr would need to be used:

```suggestion
  char *end;
  if(!buf)
    return NULL;
  end = memchr(src, '\0', length);
  if(end)
    length = end - src;
```

Reply:
Because nowhere in the code does it assume that it should only add the substring up to a null terminator. They all want the specified subtring allocated, with a null terminator so that it can be treated as a string. 

In some cases the substring might contain a null byte and then the added null terminator is of course unnecessary but that's how the code already works. If the memchr() logic remains , this use case breaks.

Reply:
Can you give me an example? I still don't get it.